app/services/data_requests/lookup_shared_device_users.rb

```diff
@@ -0,0 +1,71 @@
+module DataRequests
+  ##
+  # This class recursively looks up networks of users who are sharing devices.
+  # It does this by first looking up the devices for a set of users.
+  # Then it looks at all of the users who have signed in from those devices.
+  # If a user who has not had their devices looked up is present, it looks up
+  # the devices for that user.
+  # This process is done recuresively until all devices and users within a
+  # network are found.
+  #
+  class LookupSharedDeviceUsers
+    attr_reader :initial_users
+
+    def initialize(initial_users)
+      @initial_users = initial_users
+    end
+
+    def call
+      lookup_users(initial_users)
+      user_uuids_by_device
+    end
+
+    private
+
+    def lookup_device(device)
+      cookie_uuid = device.cookie_uuid
+      return if user_uuids_by_device.keys.include?(cookie_uuid)
+
+      warn "Searching for new devices matching #{cookie_uuid}"
+      matching_device_records = Device.includes(:user).where(cookie_uuid: cookie_uuid).all
+      user_uuids_by_device[cookie_uuid] = matching_device_records.map(&:user).compact.map(&:uuid)
```

Comment:
Since we just want the `uuid` attribute, we can use `.pluck` and `.joins` to get it. This should be faster because it's only allocating an array of primitives, instead of all the User and Device record instances.

```suggestion
      user_uuids = matching_device_records.map(&:user).compact.map(&:uuid)
      user_uuids_by_device[cookie_uuid] = user_uuids
```

Reply:
```
User.joins(:devices).where(devices: { cookie_uuid: cookie_uuid }).pluck(:uuid)
```